### PR TITLE
fix(dc) exclude cases where "plugins" are used not as plugins

### LINKED ISSUE #7160: In the use of declarative configuration , custom Dao contains a field named "plugins", which can cause startup exceptions
### Summary

In the use of declarative configuration , custom Dao contains a field named "plugins", which can cause startup exceptions

### Steps To Reproduce

1. Create plugin and define Daos, which are defined as follows:
 local typedefs = require "kong.db.schema.typedefs"
return { demo = {
  name = "demo",
  primary_key = { "id" },
  admin_api_name = "demo",
  admin_api_nested_name = "demo",
  fields = {
    { id = typedefs.uuid },
    { version = { type = "string", required = true } },
    **{ plugins = { type = "string", required = true } }**
  }
} } 

3. start kong
docker run -ti \
--name kong \
-e KONG_DATABASE=off \
-e KONG_PLUGINS=demo \
-e KONG_DECLARATIVE_CONFIG=/etc/kong/kong.yml \
-v /opt/test/kong.yml:/etc/kong/kong.yml \
-v /opt/test/demo:/usr/local/share/lua/5.1/kong/plugins/demo \
kong/kong-gateway:2.3.3.2-alpine

4. exception as follow:
2021/05/17 14:05:55 [error] 1#0: init_by_lua error: /usr/local/share/lua/5.1/kong/db/schema/init.lua:2197: attempt to index local 'self' (a nil value)
stack traceback:
        /usr/local/share/lua/5.1/kong/db/schema/init.lua:2197: in function 'new_subschema'
        /usr/local/share/lua/5.1/kong/db/schema/plugin_loader.lua:221: in function 'load_subschema'
        ...are/lua/5.1/kong/db/schema/others/declarative_config.lua:255: in function 'load_plugin_subschemas'
        ...are/lua/5.1/kong/db/schema/others/declarative_config.lua:267: in function 'load_plugin_subschemas'
        ...are/lua/5.1/kong/db/schema/others/declarative_config.lua:783: in function 'load'
        /usr/local/share/lua/5.1/kong/db/declarative/init.lua:51: in function 'new_config'
        /usr/local/share/lua/5.1/kong/init.lua:351: in function 'parse_declarative_config'
        /usr/local/share/lua/5.1/kong/init.lua:571: in function 'init'
        init_by_lua:3: in main chunk
nginx: [error] init_by_lua error: /usr/local/share/lua/5.1/kong/db/schema/init.lua:2197: attempt to index local 'self' (a nil value)
stack traceback:
        /usr/local/share/lua/5.1/kong/db/schema/init.lua:2197: in function 'new_subschema'
        /usr/local/share/lua/5.1/kong/db/schema/plugin_loader.lua:221: in function 'load_subschema'
        ...are/lua/5.1/kong/db/schema/others/declarative_config.lua:255: in function 'load_plugin_subschemas'
        ...are/lua/5.1/kong/db/schema/others/declarative_config.lua:267: in function 'load_plugin_subschemas'
        ...are/lua/5.1/kong/db/schema/others/declarative_config.lua:783: in function 'load'
        /usr/local/share/lua/5.1/kong/db/declarative/init.lua:51: in function 'new_config'
        /usr/local/share/lua/5.1/kong/init.lua:351: in function 'parse_declarative_config'
        /usr/local/share/lua/5.1/kong/init.lua:571: in function 'init'
        init_by_lua:3: in main chunk

### Additional Details & Logs

- Kong version : 1.5.1 and 2.3.3.2-alpine
- Operating system : centos7 

At present, my solution is to add verification logic to distinguish between plugins schema and custom "plugins" field. The details are as follows:
https://github.com/Kong/kong/blob/master/kong/db/schema/others/declarative_config.lua
The original code:
image
Code I changed:
 
...
 if fname == "plugins" **and fdata.elements and fdata.elements.subschema_key** then
     for plugin in pairs(plugin_set) do
...

Do not confirm whether it is the most appropriate solution, please advise?

### LINKED ISSUE #7610: "RBAC is disabled!" and "Authentication is not enabled." messages regardless of correct auth configuration
### Summary
- Kong version: Kong Gateway Enterprise 2.4.1.1
- Despite of following the documentation and correct (afaik) configuration of authentication the Kong Manager still reports that RBAC is disabled and that Authentication is not enabled.

### Steps To Reproduce

1. I installed Kong Gateway Enterprise 2.4.1.1 using these instructions: https://docs.konghq.com/enterprise/2.4.x/deployment/installation/amazon-linux-2/
2. I then followed to https://docs.konghq.com/enterprise/2.4.x/kong-manager/security/ and https://docs.konghq.com/enterprise/2.4.x/kong-manager/authentication/basic/ and https://docs.konghq.com/enterprise/2.4.x/kong-manager/authentication/super-admin/ to configure security.
3. Despite of following this rather simple set of instructions carefuly, I still receive the "RBAC is disabled! Configuration will not be applied until RBAC is enabled." message on `/teams` page.
4. I also tried `/login` page but the message says "Authentication is not enabled."
5. When visiting the <host>:8001 (kong admin API root) I noticed that fields like `rbac` and `enforce_rbac` are set to `"off"` even though they are 100% set to `on` in `kong.conf` and all the relevant configuration is done according to documentation.

There is nothing useful or at least obvious in startup or error logs.

What am I missing? How can I further debug or troubleshoot this? This seems like a bug to me, hence the bug report, but maybe I'm just so blind. Also noticed very similar issue reported in the Charts repository: https://github.com/Kong/charts/issues/357 - maybe these are related.

### Additional Details & Logs
- Operating system: Amazon Linux 2 (AWS)

### PR DESCRIPTION
**Pull request by fffonion**
entities

Fix #7160

### CODE REVIEW

File: kong/db/schema/others/declarative_config.lua [hunk 1]
```diff
@@ -243,14 +243,12 @@ local function load_plugin_subschemas(fields, plugin_set, indent)
   for _, f in ipairs(fields) do
     local fname, fdata = next(f)
 
-    if fname == "plugins" then
+    -- Exclude cases where `plugins` are used expect from plugins entities.
+    -- This assumes other entities doesn't have `name` as its subschema_key.
+    if fname == "plugins" and fdata.elements and fdata.elements.subschema_key == "name" then
       for plugin in pairs(plugin_set) do
-        local _, err, err_t = plugin_loader.load_subschema(fdata.elements, plugin, errors)
+        local _, err = plugin_loader.load_subschema(fdata.elements, plugin, errors)
```
Comment (gszr) [inline kong/db/schema/others/declarative_config.lua:250]: Missing error handling?

**Review by gszr — COMMENTED** [review]
Nice one! Could we add a test for this change?

File: kong/db/schema/others/declarative_config.lua (hunk 1 above)
Comment (fffonion) [reply kong/db/schema/others/declarative_config.lua:250]: I think plugin_loader code only returns "ok, err" not third argument, so I removed them for clearity.

**Comment by fffonion** [comment]
@gszr test added!

File: kong/db/schema/others/declarative_config.lua (hunk 1 above)
Comment (gszr) [reply kong/db/schema/others/declarative_config.lua:250]: oops, I missed the handling below 😆 apologies

**Review by gszr — APPROVED** [review]

**Pull request merged**